dask/array/core.py

```diff
@@ -1379,6 +1380,23 @@ def __getitem__(self, index):
         if all(isinstance(i, slice) and i == slice(None) for i in index2):
             return self
 
+        for axis, ind in enumerate(index2):
+            if isinstance(ind, np.ndarray):
+                cum_chunks = np.cumsum(self.chunks[axis])
+                chunk_locations = np.searchsorted(cum_chunks, ind, side='right')
+                where = np.where(np.diff(chunk_locations))[0] + 1
+
+                if len(where) > len(self.chunks[axis]) ** 2:
+                    ind1, ind2, ind3 = shuffle_convert(ind, self.chunks, axis)
+                    index3 = index2[:axis] + (ind2,) + index2[axis + 1:]
+                    index4 = tuple([slice(None, None) for ax in index2[:axis] if not
+                            isinstance(ax, int)]) + (ind3,)
+                    index4 = (slice(None, None),) * index2[:axis].count(None) + index4
+                    locs = np.concatenate([[0], np.where(ind2[1:] <
+                        ind2[:-1])[0] + 1, [len(ind2)]])
+                    c = tuple(np.diff(locs).tolist())
+                    return self[index3].rechunk({axis: c})[index4]
```

Comment:
This whole block is quite complicated, and could use at least some block-level comments to explain what each operation is doing.